File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/creator_inbox/CreatorInboxViewModel.kt
@file:Suppress("PackageNaming", "MagicNumber", "TooManyFunctions")

package app.pantopus.android.ui.screens.creator_inbox

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import app.pantopus.android.data.api.models.audience.PersonaThreadDto
import app.pantopus.android.data.api.net.NetworkResult
import app.pantopus.android.data.audience.AudienceProfileRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import java.time.Duration
import java.time.Instant
import javax.inject.Inject

/**
 * Backs the P1.2 Creator Inbox screen. Loads the owner persona +
 * followers DM threads from the same endpoints the Audience Profile
 * Threads tab uses (`/api/personas/me` + `/api/personas/:id/dms/threads`)
 * and projects them into filter-aware row models. Filter chip counts
 * derive from the loaded thread list so they always match what the
 * user sees.
 */
@HiltViewModel
class CreatorInboxViewModel
    @Inject
    constructor(
        private val repository: AudienceProfileRepository,
    ) : ViewModel() {
        private val _state = MutableStateFlow<CreatorInboxUiState>(CreatorInboxUiState.Loading)
        val state: StateFlow<CreatorInboxUiState> = _state.asStateFlow()

        private val _activeFilter = MutableStateFlow(CreatorInboxFilter.All)
        val activeFilter: StateFlow<CreatorInboxFilter> = _activeFilter.asStateFlow()

        private var threads: List<PersonaThreadDto> = emptyList()
        private var header: CreatorInboxHeader =
            CreatorInboxHeader(title = "Creator inbox", handle = null, isCrossPersona = false)

        fun load() {
            _state.value = CreatorInboxUiState.Loading
            viewModelScope.launch {
                runCatching { fetchAndProject() }
                    .onFailure {
                        _state.value = CreatorInboxUiState.Error("Couldn't load your inbox.")
                    }
            }
        }

        fun refresh() = load()

        fun selectFilter(filter: CreatorInboxFilter) {
            _activeFilter.value = filter
            rebuild()
        }

        /** Resolve a thread row's counterparty for the chat conversation
         *  push — prefer the explicit `counterpartyUserId`, fall back to
         *  the row id (server defaults that to the membership id today). */
        fun conversationPeer(row: CreatorInboxRowContent): ConversationPeer =
            ConversationPeer(
                userId = row.counterpartyUserId ?: row.id,
                displayName = row.displayName.ifEmpty { row.handle },
                initials = row.initials,
                verified = row.verifiedLocal,
            )

        private suspend fun fetchAndProject() {
            val me =
                when (val r = repository.me()) {
                    is NetworkResult.Success -> r.data
                    is NetworkResult.Failure -> {
                        _state.value = CreatorInboxUiState.Error("Couldn't load your inbox.")
                        return
                    }
                }
            val persona = me.persona
            val handle = persona?.handle
            if (persona == null || handle == null) {
                val emptyHeader =
                    CreatorInboxHeader(title = "Creator inbox", handle = null, isCrossPersona = false)
                header = emptyHeader
                _state.value = CreatorInboxUiState.Empty(emptyHeader)
                return
            }
            header =
                CreatorInboxHeader(
                    title = "Creator inbox",
                    handle = "@$handle",
                    isCrossPersona = false,
                )
            val resp =
                when (val r = repository.threads(persona.id)) {
                    is NetworkResult.Success -> r.data
                    is NetworkResult.Failure -> {
                        _state.value = CreatorInboxUiState.Error("Couldn't load your inbox.")
                        return
                    }
                }
            threads = resp.threads
            rebuild()
        }

        private fun rebuild() {
            if (threads.isEmpty()) {
                _state.value = CreatorInboxUiState.Empty(header)
                return
            }
            val rows = threads.map { row(it) }
            val counts =
                CreatorInboxCounts(
                    total = rows.size,
                    unread = rows.count { it.unread },
                    flagged = rows.count { it.flagged },
                )
            val chips = chips(rows = rows, counts = counts)
            val filter = _activeFilter.value
            val filtered = rows.filter { matches(it, filter) }
            _state.value =
                CreatorInboxUiState.Loaded(
                    CreatorInboxLoaded(
                        header = header,
                        rows = filtered,
                        counts = counts,
                        chips = chips,
                    ),
                )
        }

        companion object {
            internal fun chips(
                rows: List<CreatorInboxRowContent>,
                counts: CreatorInboxCounts,
            ): List<CreatorInboxChipContent> {
                val bronzePlus = rows.count { it.tierRank >= 2 }
                return listOf(
                    CreatorInboxChipContent(
                        id = CreatorInboxFilter.All.key,
                        filter = CreatorInboxFilter.All,
                        label = CreatorInboxFilter.All.title,
                        count = counts.total,
                    ),
                    CreatorInboxChipContent(
                        id = CreatorInboxFilter.Unread.key,
                        filter = CreatorInboxFilter.Unread,
                        label = CreatorInboxFilter.Unread.title,
                        count = counts.unread,
                    ),
                    CreatorInboxChipContent(
                        id = CreatorInboxFilter.BronzePlus.key,
                        filter = CreatorInboxFilter.BronzePlus,
                        label = CreatorInboxFilter.BronzePlus.title,
                        count = bronzePlus,
                    ),
                    CreatorInboxChipContent(
                        id = CreatorInboxFilter.Flagged.key,
                        filter = CreatorInboxFilter.Flagged,
                        label = CreatorInboxFilter.Flagged.title,
                        count = counts.flagged,
                    ),
                )
            }

            internal fun matches(
                row: CreatorInboxRowContent,
                filter: CreatorInboxFilter,
            ): Boolean =
                when (filter) {
                    CreatorInboxFilter.All -> true
                    CreatorInboxFilter.Unread -> row.unread
                    CreatorInboxFilter.BronzePlus -> row.tierRank >= 2
                    CreatorInboxFilter.Flagged -> row.flagged
                }

            internal fun row(dto: PersonaThreadDto): CreatorInboxRowContent {
                val handle = dto.fanHandle.orEmpty()
                val displayName = dto.fanDisplayName ?: handle.ifEmpty { "Follower" }
                return CreatorInboxRowContent(
                    id = dto.id,
                    displayName = displayName,
                    handle = if (handle.isEmpty()) "" else "@$handle",
                    initials = initials(displayName, handle),
                    avatarUrl = dto.fanAvatarUrl,
                    tierName = dto.tier?.name,
                    tierRank = dto.tier?.rank ?: 1,
                    preview = dto.lastMessagePreview.orEmpty(),
                    timeAgo = timeAgo(dto.lastMessageAt),
                    unread = (dto.unreadCount ?: 0) > 0,
                    flagged = dto.flagged ?: false,
                    verifiedLocal = dto.verifiedLocal ?: false,
                    counterpartyUserId = dto.counterpartyUserId,
                    personaChip = null,
                )
            }

            internal fun initials(
                name: String,
                handle: String,
            ): String {
                val source = name.ifEmpty { handle }
                val parts = source.split(" ").filter { it.isNotEmpty() }.take(2)
                val letters = parts.mapNotNull { it.firstOrNull()?.toString() }.joinToString("").uppercase()
                if (letters.isNotEmpty()) return letters
                return source.take(2).uppercase()
            }

            internal fun timeAgo(iso: String?): String {
                if (iso.isNullOrBlank()) return ""
                val instant = runCatching { Instant.parse(iso) }.getOrNull() ?: return ""
                val secs = Duration.between(instant, Instant.now()).seconds.coerceAtLeast(0)
                val mins = secs / 60
                if (mins < 1) return "Just now"
                if (mins < 60) return "${mins}m"
                val hrs = mins / 60
                if (hrs < 24) return "${hrs}h"
                val days = hrs / 24
                if (days < 7) return if (days == 1L) "Yesterday" else "${days}d"
                return "${days / 7}w"
            }
        }

        /** Counterparty payload handed back to the host so it can build
         *  the existing chat-conversation route without re-introspecting
         *  the row. Mirrors iOS `CreatorInboxConversationDestination`. */
        data class ConversationPeer(
            val userId: String,
            val displayName: String,
            val initials: String,
            val verified: Boolean,
        )
    }

File: frontend/apps/ios/Pantopus/Features/CreatorInbox/CreatorInboxViewModel.swift
//
//  CreatorInboxViewModel.swift
//  Pantopus
//
//  Backs the P1.2 Creator Inbox screen. Loads the owner persona +
//  followers DM threads from the same endpoints the Audience Profile
//  Threads tab uses (`/api/personas/me` + `/api/personas/:id/dms/threads`)
//  and projects them into filter-aware row models. Filter chip counts
//  derive from the loaded thread list so they always match what the
//  user sees.
//

import Foundation
import Observation

@Observable
@MainActor
public final class CreatorInboxViewModel {
    public private(set) var state: CreatorInboxState = .loading
    public var activeFilter: CreatorInboxFilter = .all

    private let api: APIClient
    private var threads: [PersonaThreadDTO] = []
    private var header = CreatorInboxHeader(
        title: "Creator Inbox",
        handle: nil,
        isCrossPersona: false
    )

    public init() {
        api = .shared
    }

    init(api: APIClient) {
        self.api = api
    }

    public func load() async {
        state = .loading
        do {
            let me: PersonaMeResponse = try await api.request(AudienceProfileEndpoints.me)
            guard let persona = me.persona, let handle = persona.handle else {
                let emptyHeader = CreatorInboxHeader(
                    title: "Creator Inbox",
                    handle: nil,
                    isCrossPersona: false
                )
                header = emptyHeader
                state = .empty(header: emptyHeader)
                return
            }
            let resolvedHeader = CreatorInboxHeader(
                title: "Creator Inbox",
                handle: "@\(handle)",
                isCrossPersona: false
            )
            header = resolvedHeader
            let response: PersonaThreadsResponse =
                try await api.request(AudienceProfileEndpoints.threads(personaId: persona.id))
            threads = response.threads
            rebuild()
        } catch {
            let message = (error as? APIError)?.errorDescription ?? "Couldn't load your inbox."
            state = .error(message: message)
        }
    }

    public func refresh() async {
        await load()
    }

    public func selectFilter(_ filter: CreatorInboxFilter) {
        activeFilter = filter
        rebuild()
    }

    /// Resolve a thread row's counterparty for the `ChatConversationView`
    /// push — prefer the explicit `counterpartyUserId`, fall back to the
    /// row id (server defaults that to the membership id today).
    public func conversationDestination(for row: CreatorInboxRowContent) -> CreatorInboxConversationDestination {
        let userId = row.counterpartyUserId ?? row.id
        return CreatorInboxConversationDestination(
            userId: userId,
            displayName: row.displayName.isEmpty ? row.handle : row.displayName,
            initials: row.initials,
            verified: row.verifiedLocal
        )
    }

    // MARK: - Projection

    private func rebuild() {
        if threads.isEmpty {
            state = .empty(header: header)
            return
        }
        let rows = threads.compactMap(Self.row)
        let counts = CreatorInboxCounts(
            total: rows.count,
            unread: rows.filter(\.unread).count,
            flagged: rows.filter(\.flagged).count
        )
        let chips = Self.chips(rows: rows, counts: counts)
        let filtered = rows.filter { Self.matches($0, filter: activeFilter) }
        let loaded = CreatorInboxLoaded(
            header: header,
            rows: filtered,
            counts: counts,
            chips: chips
        )
        state = .loaded(loaded)
    }

    static func chips(rows: [CreatorInboxRowContent], counts: CreatorInboxCounts) -> [CreatorInboxChipContent] {
        let bronzePlus = rows.filter { $0.tierRank >= 2 }.count
        return [
            CreatorInboxChipContent(filter: .all, count: counts.total),
            CreatorInboxChipContent(filter: .unread, count: counts.unread),
            CreatorInboxChipContent(filter: .bronzePlus, count: bronzePlus),
            CreatorInboxChipContent(filter: .flagged, count: counts.flagged)
        ]
    }

    static func matches(_ row: CreatorInboxRowContent, filter: CreatorInboxFilter) -> Bool {
        switch filter {
        case .all: true
        case .unread: row.unread
        case .bronzePlus: row.tierRank >= 2
        case .flagged: row.flagged
        }
    }

    static func row(_ dto: PersonaThreadDTO) -> CreatorInboxRowContent? {
        let handle = dto.fanHandle ?? ""
        let displayName = dto.fanDisplayName ?? (handle.isEmpty ? "Follower" : handle)
        let initials = initials(of: displayName, handle: handle)
        return CreatorInboxRowContent(
            id: dto.id,
            displayName: displayName,
            handle: handle.isEmpty ? "" : "@\(handle)",
            initials: initials,
            avatarUrl: dto.fanAvatarUrl,
            tierName: dto.tier?.name,
            tierRank: dto.tier?.rank ?? 1,
            preview: dto.lastMessagePreview ?? "",
            timeAgo: timeAgo(from: dto.lastMessageAt),
            unread: (dto.unreadCount ?? 0) > 0,
            flagged: dto.flagged ?? false,
            verifiedLocal: dto.verifiedLocal ?? false,
            counterpartyUserId: dto.counterpartyUserId,
            personaChip: nil
        )
    }

    static func initials(of name: String, handle: String) -> String {
        let source = name.isEmpty ? handle : name
        let parts = source.split(separator: " ").prefix(2)
        let letters = parts.compactMap { $0.first.map(String.init) }.joined().uppercased()
        if !letters.isEmpty { return letters }
        return String(source.prefix(2)).uppercased()
    }

    static func timeAgo(from iso: String?) -> String {
        guard let iso, let date = ISO8601DateFormatter().date(from: iso) else { return "" }
        let interval = Date().timeIntervalSince(date)
        let minutes = Int(interval / 60)
        if minutes < 1 { return "Just now" }
        if minutes < 60 { return "\(minutes)m" }
        let hours = minutes / 60
        if hours < 24 { return "\(hours)h" }
        let days = hours / 24
        if days < 7 { return days == 1 ? "Yesterday" : "\(days)d" }
        let weeks = days / 7
        return "\(weeks)w"
    }
}
